### DISCUSSION TITLE
How to preprocess row before deserializing to struct?

### DISCUSSION THREAD

**Discussion by evbo** [BurntSushi/rust-csv#367] [Q&A]
Hi, I have a disgusting csv format where someone thought it was clever to omit delimiters when the remaining columns of a row are all null.

So for example:
```
x,y,z,
1,2,3,
1
```

instead of :
```
x,y,z,
1,2,3,
1,,,
```

What I would like to do is append as many `,` as are missing to the `StringRecord` and then ask `csv` to deserialize it into a struct:

```rust
struct Xyz {
  x: i32,
  y: Option<i32>,
  z: Option<i32>
}
```

I know how to append to the `as_slice()` result from `StringRecord`. but how can I then ask this library to deserialize the `&str` after I've appended the missing `,`?

**Comment by BurntSushi** [comment]
Once you have a `StringRecord`, you can use its [`StringRecord::deserialize`](https://docs.rs/csv/latest/csv/struct.StringRecord.html#method.deserialize) method to invoke serde. Or just do it by hand if your fields are just numbers.

**Reply by evbo** [reply]
I have a `String` (after appending `,`), not a `StringRecord`, but I presume I can convert a `String` into `StringRecord`?

**Reply by BurntSushi** [reply]
You need to use a `csv::Reader` to parse a `&str` into a `StringRecord`.

**Reply by evbo** [reply]
hmmm, but then that means the header information is lost since it's no longer the same reader I was initially using?

So would it make more sense to just loop through the file twice?:

1. to add missing delimiters
2. to deserialize each row with the header information

**Reply by BurntSushi** [reply]
You can set the headers: https://docs.rs/csv/latest/csv/struct.Reader.html#method.set_headers

It indeed might be simpler to just do two passes. It may or may not be slower though.

**Reply by evbo** [reply]
I opted for 2 passes as it helps separate concerns too.

However, I was hoping `as_slice` could be configured to include the delimiter. Would you consider such an enhancement for those of with broken csv that needs monkey patching?

Otherwise, it's pretty cumbersome just getting a string version of the record:

```rust
let mut wtr = Writer::from_writer(vec![]).delimiter(b'\t');
wtr.write_record(string_record.into_iter())?;
let string_data = String::from_utf8(wtr.into_inner()?)?;

// dream alternative (could work with BytesRecord to produce slice of delimited bytes too):
let string_data = string_record.serialized() // e.g. produces: "1,2,3,"
```

**Reply by BurntSushi** [reply]
No. It doesn't really make sense to put the delimiter in there.

If you really need precise control, you might need to use `csv-core`.

**Reply by evbo** [reply]
looks like `csv-core` is for `no_std`... not sure I want to delve that deeply quite yet or is my use of `write_record` above not recommended?

**Reply by BurntSushi** [reply]
It's odd I guess. Honestly, you have an odd use case, and I think you're just not going to have a nice paved path here. 

I think if I were in your shoes I would have just written a custom parser for my data-that-looks-like-csv-but-isn't.

**Reply by evbo** [reply]
one last suggestion if it's okay: `serde` handles this scenario by presuming a missing value should be `None`:
```rust
#[serde(deserialize_with = "object_empty_as_none")]
```

Do you see supporting that same feature for missing csv columns being possibly valuable? I did learn the other day Excel/Libreoffice will load csv with inconsistent row lengths (missing delimiters)

**Reply by BurntSushi** [reply]
Have you tried ReaderBuilder::flexible? If so, what happens?

**Reply by evbo** [reply]
yes `flexible` is required for it to read this sort of csv. I mean the step involving deserializing could support missing delimiters as assuming they are `None` values for those columns missing.

**Reply by BurntSushi** [reply]
Maybe. I'm not sure if the serde deserializer for CSV can easily do it. You're welcome to try submitting a patch for it. But the code is extremely hairy. It's a big part of what this crate hasn't evolved much and why there are so many serde caveats. Look at the issue tracker. There are tons of issues and almost all of them are related to serde. But changing the deserializer is just something that I've personally found very difficult.

**Comment by korrat** [comment]
I'm working on a similarly bad format. What I do is pad the record just before deserialization: 

```rust
record.extend(core::iter::repeat("").take(headers.len() - record.len()));
let record = match record.deserialize(Some(headers));
```
